Provide the response and error message as write-out "variables"

fearblackcat opened the pull request:
This pull request is to fix the stdout and stderr in write out option. You can see my completed effect in below illustration:
![QQ图片20190720213352](https://user-images.githubusercontent.com/5964503/61579311-2e9aee80-ab36-11e9-90ba-956f04e2443e.png)
![QQ图片20190720213407](https://user-images.githubusercontent.com/5964503/61579312-3195df00-ab36-11e9-8520-074a029a47f5.png)

Notice:
The curl-format.t content is 
``
{"http_code" : %{http_code}, "time_namelookup":  %{time_namelookup}, "time_connect":  %{time_connect}, "time_appconnect": %{time_appconnect},"time_pretransfer": %{time_pretransfer}, "time_redirect": %{time_redirect}, "time_starttransfer": %{time_starttransfer}, "time_total": %{time_total}, "stdout": "%{stdout}", "stderr": "%{stderr}"}
``

bagder commented:
Please elaborate more on the change. What does this change and why. This pr even adds public APIs and other public symbols and we do not do that without careful consideration!

fearblackcat commented:
> Please elaborate more on the change. What does this change and why. This pr even adds public APIs and other public symbols and we do not do that without careful consideration!

@bagder  my purpose is put the response content or stderr into a buffer. But the implementation of curl use the file handle for the output. (the original code assign the stdout or stderr to the FILE* if not specify the file download).  I need a buffer bind to file handle. Therefore I use the open_memstream to encapsulate encapsulate two public methods. 

Notice:
1. I also occurs some problem when I used the glibc api.  It's must be 10.13 version for macos so I changed the build flag in acinclude.m4.  
2. It cannot affect when use the write out option stdout and stderr before I changed code. The response content cannot assign to the stdout option cause the whole output content is stdout.

bagder commented:
> @bagder my purpose is put the response content or stderr into a buffer.

The response *content*?  It can already get put into a buffer fine, see the [getinmemory.c](https://curl.haxx.se/libcurl/c/getinmemory.html) example for one way to do it. No change at all is needed for that...

fearblackcat commented:
> > @bagder my purpose is put the response content or stderr into a buffer.
> 
> The response _content_? It can already get put into a buffer fine, see the [getinmemory.c](https://curl.haxx.se/libcurl/c/getinmemory.html) example for one way to do it. No change at all is needed for that...

This example cannot solve my problem. I need the bind the file handle with buffer cause the output data is redirect to the file stream.
![QQ截图20190721114221](https://user-images.githubusercontent.com/5964503/61586663-a572ce80-abac-11e9-986d-76b4f591125c.png)

bagder commented:
> This example cannot solve my problem.

Which problem is that?

> I need bind the file handle with buffer and prevent the original response to output in stdout or stderr cause the output data is redirect to the file stream.

libcurl provides callbacks and user-provided pointers for all delivered data, it doesn't have to use any files or streams if you prefer something else.

fearblackcat commented:
> libcurl provides callbacks and user-provided pointers for all delivered data, it doesn't have to use any files or streams if you prefer something else.

I understand ur meanings. But this callbacks have already registered and  used for all of complicated scenes.
![QQ截图20190721201623](https://user-images.githubusercontent.com/5964503/61591111-dbd43c00-abf4-11e9-8ff3-c64bdd5e2a9a.png)
The above image shows in operate_do function.

bagder commented:
> But these callbacks have already registered and used for all of complicated scenes.

You still haven't explained what problem you're trying to solve with this effort so I can't really suggest a solution either.

(Adding a new callback to libcurl for the purpose of using in the curl tool simply because the existing functionality done when using the existing callbacks is complicated seems like a really bad excuse.)

fearblackcat commented:
> You still haven't explained what problem you're trying to solve with this effort so I can't really suggest a solution either.

@bagder  As you can see the below image. The output content is not correspond with the value of stdout position. There is wide difference to my first comment as before.

![QQ图片20190722022008](https://user-images.githubusercontent.com/5964503/61595321-333fcf80-ac28-11e9-8a4c-72b98681965a.png)

bagder commented:
In the first invoke you get both content (response) and write-out date sent to stdout. **since you asked for it**

In the second invoke you seem to not get any response output at all **even though the command line asked for it**.

So no, I don't see any fix. I see a broken curl. Care you explain what the fix is?

fearblackcat commented:
> In the first invoke you get both content (response) and write-out data sent to stdout. **since you asked for it**
> 
> In the second invoke you seem to not get any response output at all **even though the command line asked for it**.

@bagder  The second case is I intentionally disconnected the network to see if the error message was in the value position of json sterr. 
I just want something like this:
![61579311-2e9aee80-ab36-11e9-90ba-956f04e2443e](https://user-images.githubusercontent.com/5964503/61600485-72d7dd00-ac63-11e9-8b20-1fab79b06221.png)

![61579312-3195df00-ab36-11e9-8520-074a029a47f5](https://user-images.githubusercontent.com/5964503/61600491-79feeb00-ac63-11e9-8461-edde67a6763f.png)

bagder commented:
> Notice that the output is displayed in the corresponding value location of json

You mean that the error message is after the response content? But the error message is to stderr and the content is to stdout, right? So again: what exactly does your fix do?

fearblackcat commented:
> You mean that the error message is after the response content? But the error message is to stderr and the content is to stdout, right? So again: what exactly does your fix do?

That was not what I meant. The two screenshots above are not related, just two test cases. The two screenshots above correspond to the two screenshots of the previous comments.

You should look carefully between the bellow images.
**case 1:**
***before fixed***
![61595321-333fcf80-ac28-11e9-8a4c-72b98681965a](https://user-images.githubusercontent.com/5964503/61627242-d983e780-acb1-11e9-94ec-563da77c68d5.png)
***after fixed***
![61600485-72d7dd00-ac63-11e9-8b20-1fab79b06221](https://user-images.githubusercontent.com/5964503/61627271-e6a0d680-acb1-11e9-8d17-bc5fc8aa4c27.png)

**case 2:**
***before fixed***
![61595373-e6102d80-ac28-11e9-91cb-a6b9c0fe484d](https://user-images.githubusercontent.com/5964503/61627301-fcae9700-acb1-11e9-8d2a-d10646da3c9b.png)

***after fixed***
![61600491-79feeb00-ac63-11e9-8461-edde67a6763f](https://user-images.githubusercontent.com/5964503/61627322-0932ef80-acb2-11e9-9c80-027f4e52e6e1.png)

***These images are very different. I hope you can find them.**** :-(

bagder commented:
> You should look carefully between the bellow images.

You should *explain* to us what the changes are, which I think I've now asked for at least three times and you still refuse to. It looks to me like you're trying to change the order of some output and I can't understand it (or why).

fearblackcat commented:
> > You should look carefully between the bellow images.
> 
> You should _explain_ to us what the changes are, which I think I've now asked for at least three times and you still refuse to. It looks to me like you're trying to change the order of some output and I can't understand it (or why).

Not the order. The original write-out format cannot print the result value to the location of stdout and stderr. If I plan to parse the json data, I cannot get the data of sterr or stdout in json.

bagder commented:
> The original write-out format cannot print the result value to the location of stdout and stderr.

You mean write-out can't write to **both** streams or that it can't write to a specific one of them? Can you explain with more details what write-out cannot do?

fearblackcat commented:
> > The original write-out format cannot print the result value to the location of stdout and stderr.
> 
> You mean write-out can't write to **both** streams or that it can't write to a specific one of them? Can you explain with more details what write-out cannot do?

The stdout and stderr option write-out both cannot print the streams to the value location of it.
![QQ20190722-192700](https://user-images.githubusercontent.com/5964503/61629224-da6b4800-acb6-11e9-85ea-287646b85cf8.png)

Because of the stream initial value is stdout and the logic in switch case of ourWriteOut function assigned stdout or sterr to stream. It doesn't make any sense.

![QQ20190722-192925](https://user-images.githubusercontent.com/5964503/61629423-6a10f680-acb7-11e9-878d-860bf269a629.png)

gevaerts commented:
It looks like you're totally missing what %{stderr} and %{stdout} do (and are supposed to do, and are documented to do). They explicitly do *not* mean "insert whatever the remote sends here" or "insert any error message here".

Also, your changes to curl.h add duplicate enum values, and you have an entirely unrelated change to acinclude.m4.

fearblackcat commented:
> It looks like you're totally missing what %{stderr} and %{stdout} do (and are supposed to do, and are documented to do). They explicitly do _not_ mean "insert whatever the remote sends here" or "insert any error message here".
> 
> Also, your changes to curl.h add duplicate enum values, and you have an entirely unrelated change to acinclude.m4.

@gevaerts 
I think the stdout is to print the result of response content and stderr should print the error of result otherwise it make no sense.

bagder commented:
> I think the stdout is to print the result of response content and stderr should print the error of result otherwise it make no sense.

No, you cannot decide a totally new meaning for this functionality. They do exactly what they are documented to do: they change to what stream the write-out data is written to and that stream defaults to **stdout** as it always has. If you want the write-out data sent to stderr instead, which can be convenient if you want to intercept for example reponse data on stdout, you can use `%{stderr}`

Do you want to change to what stream the response data is sent to or what?

> I have already uploaded the images of differences between unchanged code result and changed code result. If you cannot understand, I can only say u r totally ignorance

As is on secret here, I have asked you many many times to explain what your change does and you have not once managed to do so. That might of course make me "ignorance" or perhaps plain stupid, but I also think it shows something else...

fearblackcat commented:
@bagder 
You don't want to merge, you can say it earlier so that you don't waste my time

bagder commented:
I can't have an positive opinion on merging until I understand the change, its reason and I agree with it (and the code is green in the CI and code review). And you're not "selling" your idea very well here since I still don't understand it...

fearblackcat commented:
The idea is to have the standard output format field with the HTTP return, the standard error field with the curl error, and the write-out format that I specify as a valid json.

I have explained it numerous times with pictures and words. But u still cannot understand. hollyshit!

bagder commented:
> The idea is to have the standard output format field with the HTTP return

It goes to stdout by default.

> the standard error field with the curl error

That goes to stderr by default

> and the write-out format that I specify as a valid json.

That too goes to stdout by default.

Which of these do you want to change? And where do you want them to go instead?

> I have explained it numerous times with pictures and words. But u still cannot understand. hollyshit!

Exactly.

bagder commented:
someone on IRC suggested this is what you want:

A `%{content}` variable for write-out that would get the response contents inserted (although you used `stdout` for that) and a separate `%{error}` that would get the error message. Is that right?

fearblackcat commented:
> > The idea is to have the standard output format field with the HTTP return
> It goes to stdout by default.
No . You can see the before and after diagram of the test results after the code changes
> > the standard error field with the curl error
> That goes to stderr by default
Not yet. This also show weird format. You also can see the before and after diagram of the test results after the code changes.

fearblackcat commented:
> someone on IRC suggested this is what you want:
> 
> A `%{content}` variable for write-out that would get the response contents inserted (although you used `stdout` for that) and a separate `%{error}` that would get the error message. Is that right?

Yes.

bagder commented:
I'm not against such an idea but the change cannot change the meaning of already existing variables, they need to be introduced using new names. Also, does the use of these variables for write-out mean that the regular response and error message outputs are silenced? It doesn't quite match how write-out works in other regards, since it is just additional output and not a behavior-changing option.

fearblackcat commented:
> I'm not against such an idea but the change cannot change the meaning of already existing variables, they need to be introduced using new names. Also, does the use of these variables for write-out mean that the regular response and error message outputs are silenced? It doesn't quite match how write-out works in other regards, since it is just additional output and not a behavior-changing option.

First of all, the person who use this format has to deal with the error message and return result by themselves, so that you can get the error message and return result in the formatted regular json, and there is no need to focus on standard output and standard error.

Similarly, don't c++ designers have to ask developers to free up memory?

bagder commented:
> there is no need to focus on standard output and standard error

 Imagine a command line like this:

    curl $URL -w @format.txt

In the file `format.txt` there's only `%{http_response}` which thus makes the HTTP response code get output on the terminal after the response body.

Or it could even look like this:

    curl $URL -o output -w @format.txt

which saves the response to the given output file.

Then, we update the format file to include `%{contents}` and rerun the same command lines. Did this now suddenly make the output not get sent to stdout and to the file name like before? I think that's a hard sell, usability wise.

> Similarly, don't c++ designers have to ask developers to free up memory?

They do. How is that relevant here?

fearblackcat commented:
> curl $URL -o output -w @format.txt
> which saves the response to the given output file.
> Then, we update the format file to include `%{contents}` and rerun the same command lines. Did this now suddenly make the output not get sent to stdout and to the file name like before? I think that's a hard sell, usability wise.

@bagder Yup, this will make the output not get sent to stdout and to the file name.  I will fix this and assign another file handle and not overwrite the original file handle.

Also I will add new write-out variables %{content} and %{error}.

Do u think this way is ok?

bagder commented:
> Do u think this way is ok?

I think the variable names could be improved but otherwise I think this concept should be fine. Do note that you closed this PR so it falls off my (and provably others') radar...

fearblackcat commented:
> I think the variable names could be improved but otherwise I think this concept should be fine. Do note that you closed this PR so it falls off my (and probably others') radar...

OK, I'll update the code.

fearblackcat commented:
@bagder  I have already updated the code.

bagder commented:
"This branch has conflicts that must be resolved" -  I suggest you rebase, squash and force-push.

Your branch still has API changes to libcurl that I disagree with. Why do you need to change the libcurl API just to add "write-out" variables to the curl tool?

fearblackcat commented:
> Your branch still has API changes to libcurl that I disagree with. Why do you need to change the libcurl API just to add "write-out" variables to the curl tool?

@bagder  I need get the buffer that point to the content and not conflict the original file stream. So that I must set the memory stream to a new file handle and assign a buffer pointer.

bagder commented:
> I need get the buffer that point to the content and not conflict the original file stream. So that I must set the memory stream to a new file handle and assign a buffer pointer.

I disagree that you *need* it.  I will not okay such a change to the libcurl API for this purpose. Please figure out a way to introduce these variables without changing the libcurl API.

fearblackcat commented:
> I disagree that you _need_ it. I will not okay such a change to the libcurl API for this purpose. Please figure out a way to introduce these variables without changing the libcurl API.

@bagder  Do you only agree to change in "src" directory?

bagder commented:
Yes, for this enhancement only `src/` files need to be changed.

fearblackcat commented:
@bagder  Please have a look.

bagder commented:
"This branch cannot be rebased due to conflicts"

Please rebase, fix the conflicts, squash into the least number of sensible commits you think it should be (that makes the PR good for reviewing), then force-push.

fearblackcat commented:
rebase and squash

fearblackcat commented:
@bagder Please help review.

bagder commented on src/tool_writeout.c:
```diff
@@ -233,13 +359,25 @@ void ourWriteOut(CURL *curl, struct OutStruct *outs, const char *writeinfo)
+                  else {
+                    fprintf(stream, "%.3f", doubleinfo);
+                  }
+                }
```
Several more unrelated changes, please remove

bagder commented on src/tool_writeout.c:
```diff
@@ -331,6 +469,24 @@ void ourWriteOut(CURL *curl, struct OutStruct *outs, const char *writeinfo)
+              case VAR_CONTENT:
+                fflush(outs->buffer_stream);
+                if(NULL != outs->bodybuf) {
+                  if('\"' == *(ptr-3)) {
```
please explain what this is? This is not documented.

is dereffing the pointer at -3 really always safe? It seems that could be before the string?

fearblackcat replied:
Already resolve.

bagder commented on src/tool_writeout.c:
```diff
@@ -105,14 +107,84 @@ static const struct variable replacements[]={
+    return NULL;
+  }
+  len = strlen(output);
+  formatStr = (char *)malloc(2*len + 1);
```
This can fail, and this code doesn't deal with that risk.

bagder commented on src/tool_writeout.c:
```diff
@@ -105,14 +107,84 @@ static const struct variable replacements[]={
+  }
+  len = strlen(output);
+  formatStr = (char *)malloc(2*len + 1);
+  memset(formatStr, 0, 2*len + 1);
```
No need to zero the string (and if you need that, consider calloc instead)

bagder commented on src/tool_writeout.c:
```diff
@@ -105,14 +107,84 @@ static const struct variable replacements[]={
 
+void copyandshift(char **replace, char *escape)
+{
+  int offset;
```
This is used to hold the *length* of the escape string, not really an offset is it? And since it holds the return from strlen() it ought to be a `size_t` and not an `int`...

bagder commented on src/tool_writeout.c:
```diff
@@ -105,14 +107,84 @@ static const struct variable replacements[]={
+    switch(*tmpbuf) {
+    case '"':
+      escapeStr = "\\\"";
+      copyandshift(&replaceStr, escapeStr);
```
couldn't these invokes then instead just be done like this to save assigning a variable that isn't really used?
~~~c
copyandshift(&replaceStr, "\\\"");

bagder commented on src/tool_writeout.c:
```diff
@@ -105,14 +107,84 @@ static const struct variable replacements[]={
+      break;
+    default:
+      if('\x00' <= *tmpbuf && *tmpbuf <= '\x1f') {
+        curlx_msprintf(replaceStr, "\\u%04x", (int)*tmpbuf);
```
So this outputs `\u00bb` for byte values below 32 and I suppose over 127 too since tmpbuf points to a signed char? What made you pick that escaping? (and it's not documented either)

fearblackcat replied:
For the shortest representation you may use JSON shortcuts, such as \" instead of \u0022. The function produces the shortest JSON representation of a UTF-8 encoded string

bagder replied:
1. JSON style is not necessarily the only way to format such bytes, so this needs to be documented.
2. There will me many cases where the data is not UTF-8.
3. Dereferencing the char pointer makes it a byte, not necessarily a full UTF-8 character surely?

fearblackcat replied:
@bagder I will documented for it.  This escape format is only for JSON format.

fearblackcat commented on acinclude.m4:
```diff
@@ -2525,7 +2525,7 @@ AC_DEFUN([CURL_MAC_CFLAGS], [
     min=""
     if test -z "$(echo $CFLAGS | grep m.*os.*-version-min)"; then
-      min="-mmacosx-version-min=10.8"
+      min="-mmacosx-version-min=10.13"
```
@bagder this change is related to open_memstream function.
If not specify the 10.13 or later, it will make failure in mac os.

fearblackcat commented on docs/cmdline-opts/write-out.d:
```diff
@@ -115,6 +115,14 @@ From this point on, the --write-out output will be written to standard output.
 (Added in 7.63.0)
 .TP
+.B content
+From this point on, the result of response content.
```
I will update this and demonstrates more details.
>(and perhaps: can we come up with a way that can offer the raw data as well as an escaped version?)
Yes, of course. If user not add quotes for %{content}. It will show the raw data otherwise display the escaped string.

fearblackcat replied:
```txt
.TP
.B response_body
The response body that the resouce requested by curl. It will be escaped by JSON
format if you add quotes for it.(eg. "%{response_body}")
(Added in 7.66.0)
.TP
.B error_message
The error message that curl returned. It will be escaped by JSON format if you
add quotes for it.(eg. "%{error_message}")
(Added in 7.66.0)

```
@bagder  Do u think this is OK?

bagder replied:
I think you should leave out the escaping part entirely in the first step and then provide a generic way to output escaped variables for write-out. I think it would be better to do that using a more explicit way than if the variable happens to have double-quotes around it. (since some users will prefer single-quotes or not using quotes at all etc).

Perhaps it could be an instruction within the braces?

    %{variable:q}

Where `q` stands for "quoted". (This would also allow for other instructions if someone has other ideas down down the line...)

fearblackcat replied:
OK, @bagder  I will adopt ur approach.

bagder commented:
There are also some [compiler errors](https://travis-ci.org/curl/curl/jobs/567741145#L2255) to address.

fearblackcat commented:
>open_memstream() is not widely available so it needs to be at least protected with #ifdef and I would generally suggest you don't use it so that this functionality can also work on windows etc.

>This of course go for both uses of this function.

There is a solution. As noted in my question, I was using open_memstream(), which is unsupported on windows.

Since I have a File * pointer (this cannot be changed to char *), I needed to redirect it to memory until later. Since I'm dealing with a file in memory, I looked into mmap(). It handily solves the problem, but again, it is linux only.

But, windows includes a corollary to mmap() called MapViewOfFile(). Through the magic of #ifdef I've got it using whichever is necessary:
```
#ifdef WIN32
#include <windows.h>
#else
#include <sys/mman.h>
#endif

```

@bagder Do u think this is ok?

bagder commented:
> Since I have a File * pointer (this cannot be changed to char *),

Why not? We get the data delivered in a callback and we can do whatever we want with it!

fearblackcat commented:
> > Since I have a File * pointer (this cannot be changed to char *),
> 
> Why not? We get the data delivered in a callback and we can do whatever we want with it!

OK. I understand it.

fearblackcat commented:
@bagder  

```c
if(!global->errors)
  global->errors = stderr;

```

The `global->errors` doesn't have callback. I need prevent the stdout and stderr to output. For the response body I can prevent output to stdout in callback. But for the error message I have no  callback function. Do u have some good advices?

bagder commented:
The error message isn't output by the library. It is written by the curl tool itself so there shouldn't be any problem to (also) store that message, or just show it again.

fearblackcat commented:
@bagder Can I change library to add a callback for error message data just like [getinmemory.c](https://curl.haxx.se/libcurl/c/getinmemory.html)?

fearblackcat commented on src/tool_operate.c:
```diff
@@ -1097,6 +1107,16 @@ static CURLcode create_transfers(struct GlobalConfig *global,
+        if(NULL != config->writeout &&
+          NULL != strstr(config->writeout, "%{error}")) {
+          FILE *errorStream = NULL;
+          errorStream = open_memstream(&outs->errorbuf, &outs->errorsize);
```
@bagder Maybe the response body can use callback function to save the content to a buffer. But the error message is possible to change the libcurl.

fearblackcat commented on src/tool_writeout.c:
```diff
@@ -105,14 +107,84 @@ static const struct variable replacements[]={
   {"stdout", VAR_STDOUT},
   {"stderr", VAR_STDERR},
+  {"content", VAR_CONTENT},
+  {"error", VAR_ERROR},
```
adopt. Thx.

fearblackcat commented:
@bagder  Can u comment my previous post?

bagder commented:
> Can I change library to add a callback for error message data

Why? The library already provides the error message, what would you gain by providing another means of getting the same thing?

fearblackcat commented:
> thing

But I also need prevent the error message output to stderr.  Such as 

```c

fprintf(global->errors, "%s\n", per->separator_err)

```
This invoke will also print error to stderr.

bagder commented:
> But I also need prevent the error message output to stderr

No you don't. The `-w` treatment is entirely separate and independent from that. If you want to silence the error output then use `-s`.

bagder commented:
Is this something you're still working on?

jay commented:
What is the use case for this for someone other than this specific user?

fearblackcat commented:
> Is this something you're still working on?

Sorry， I have other high priority work to do. But I will still working on it. I will update it.

fearblackcat commented:
> > But I also need prevent the error message output to stderr
> 
> No you don't. The `-w` treatment is entirely separate and independent from that. If you want to silence the error output then use `-s`.

@bagder  I need automatically format the error message and response data to standard json. So I   also need merge error message(include stderr) to write-out.

bagder commented:
> I need automatically format the error message and response data to standard json.

It would be really weird for curl to output one single (or two) -w variable "in json format" when *nothing else* in curl outptus anything in "json format". I can't see how that fits.

If you propose a way to format *any* or *all* the -w variables in json format with some magic, then I would be open for discussing that.

> I don't want the stderr to affect the write-out (specify standard json format).

Why would it?

fearblackcat commented:
> > I need automatically format the error message and response data to standard json.
> 
> It would be really weird for curl to output one single (or two) -w variable "in json format" when _nothing else_ in curl outptus anything in "json format". I can't see how that fits.
> 
> If you propose a way to format _any_ or _all_ the -w variables in json format with some magic, then I would be open for discussing that.
> 
> > I don't want the stderr to affect the write-out (specify standard json format).
> 
> Why would it?

```shell

cat ./curl-format.t
{"http_code" : %{http_code}, "time_namelookup":  %{time_namelookup}, "time_connect":  %{time_connect}, "time_appconnect": %{time_appconnect},"time_pretransfer": %{time_pretransfer}, "time_redirect": %{time_redirect}, "time_starttransfer": %{time_starttransfer}, "time_total": %{time_total}, "content": "%{content}", "error": "%{error}"}

```

bagder commented:
If you want curl to not display the error message, `-s` is there. Surely that's good enough?

bagder commented:
Stalled and a questionable feature.
